server/src/main/java/org/opensearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -1098,6 +1102,91 @@ private void asyncCleanupUnlinkedShardLevelBlobs(
         }
     }
 
+    public static class RemoteStoreShardCleanupTask implements Runnable {
+        private final Runnable task;
+        private final ExecutorService executor;
+        private final String shardIdentifier;
+        final static Set<String> ongoingRemoteDirectoryCleanups = newConcurrentSet();
+        final static ConcurrentMap<String, BlockingQueue<Runnable>> shardCleanupPendingTasks = new ConcurrentHashMap<>();
+
+        public RemoteStoreShardCleanupTask(Runnable task, ExecutorService executor, String indexUUID, ShardId shardId) {
+            this.task = task;
+            this.shardIdentifier = indexShardIdentifier(indexUUID, shardId);
+            this.executor = executor;
+        }
+
+        private static String indexShardIdentifier(String indexUUID, ShardId shardId) {
+            return String.join("/", indexUUID, String.valueOf(shardId.id()));
+        }
+
+        @Override
+        public void run() {
+            try {
+                if (ongoingRemoteDirectoryCleanups.add(shardIdentifier)) {
+                    task.run();
+                    Runnable nextTask = shardCleanupPendingTasks.get(shardIdentifier).poll(0L, TimeUnit.MILLISECONDS);
+                    if (nextTask != null) {
+                        executor.execute(nextTask);
+                    }
+                    ongoingRemoteDirectoryCleanups.remove(shardIdentifier);
+                } else {
+                    shardCleanupPendingTasks.putIfAbsent(shardIdentifier, new LinkedBlockingQueue<>());
```

Comment:
this `LinkedBlockingQueue` is not linked to the threadpool, so threads are unaware that there is any tasks enqueued in this which needs to be picked up. Lets say you added 4 tasks for each identifier, and currently all 4 threads of Remote purge pool, hits exception while running the tasks. The running threads will bail out and the other 3 tasks for each identifier will not get dequeued ever.

Reply:
now doing all the pending tasks of a shard in the same thread.